python/tests/ibd.py

```diff
@@ -0,0 +1,421 @@
+        """
+        The wrapper for the procedure that calculates IBD segments.
+        """
+
+        # Set up an iterator over the edges in the tree sequence.
+        mygen = iter(self.ts.edges())
+        e = next(mygen)
+        nodes_to_process = []
+
+        # Iterate over the edges.
+        while e is not None:
+            # Process all edges with the same parent node.
+            self.current_parent = e.parent
+            self.current_time = self.tables.nodes.time[self.current_parent]
+            if self.max_time is not None and self.current_time > self.max_time:
+                # Stop looking for IBD segments once the
+                # processed nodes are older than the max time.
+                break
+
+            seg = Segment(e.left, e.right, e.child)
+
+            # Create a SegmentList() holding all segments that descend from seg.
+            list_to_add = SegmentList()
+            u = seg.node
+            if u in self.samples:
+                list_to_add.add(seg)
+            else:
+                if self.A[u] is not None:
+                    s = self.A[u].head
+                    while s is not None:
+                        intvl = (
+                            max(seg.left, s.left),
+                            min(seg.right, s.right),
+                        )
+                        if intvl[1] - intvl[0] > 0:
+                            list_to_add.add(Segment(intvl[0], intvl[1], s.node))
+                        s = s.next
+
+            pn = self.calculate_ibd_segs(self.current_parent, list_to_add)
+            nodes_to_process += pn
+
+            # For parents that are also samples
+            if (self.current_parent in self.samples) and self.PARENT_SHOULD_BE_ADDED:
+                singleton_seg = SegmentList()
+                singleton_seg.add(
+                    Segment(0, self.ts.sequence_length, self.current_parent)
+                )
+                pn = self.calculate_ibd_segs(self.current_parent, singleton_seg)
+                self.PARENT_SHOULD_BE_ADDED = False
+                nodes_to_process += pn
+
+            # Move to next edge.
+            e = next(mygen, None)
+
+            # Remove any processed nodes that are no longer needed.
+            if e is None or e.parent != self.current_parent:
+                nodes_to_remove = list(set(nodes_to_process))
+                for n in nodes_to_remove:
+                    if self.current_parent in self.parent_list[n]:
+                        self.parent_list[n].remove(self.current_parent)
+                    if len(self.parent_list[n]) == 0:
+                        self.A[n] = []
+                pn = []
+
+        return self.ibd_segments
+
+    def calculate_ibd_segs(self, current_parent, list_to_add):
+        """
+        Write later.
+        """
+
+        if self.A[current_parent] is None:
+            self.A[current_parent] = list_to_add
+
+        else:
+            seg0 = self.A[current_parent].head
```

Comment:
This is essentially doing
```
for seg0 in self.A[current_parent]:
  for seg1 in list_to_add:
     ...
```
right? it'd be a bit more clear written that way...

Reply:
The reason why I haven't done this is that `self.A[current_parent]` and `list_to_add` are `SegmentList` objects rather than actual lists. So you can't cycle through them using `for`, you have to explicitly use the `next` attribute to get each segment in the sequence